### PR TITLE
INSP: Extend function argument count check to work for closures (E0057)

### PR DESCRIPTION
**Pull request by zeroeightysix**
This PR extends the inspection for fixing an incorrect amount of arguments passed to a function to work for closures as well. This corresponds to rust error [E0057](https://doc.rust-lang.org/error-index.html#E0057).

This PR might clash with #6318 as both modify the same enum, on the same line.

Part of #886

changelog: Detect [E0057](https://doc.rust-lang.org/error-index.html#E0057) in some cases

### CODE REVIEW

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 1]
```diff
@@ -1257,6 +1263,15 @@ private fun RsCallExpr.expectedParamsCount(): Pair<Int, Boolean>? {
             val s = if (el.selfParameter != null) 1 else 0
             Pair(count + s, el.isVariadic)
         }
+        is RsPatBinding -> {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:1266]: Note, I can write
```rust
struct S(&'static dyn Fn(i32));

const F: &'static dyn Fn(i32) = &bar;

fn main() {
    let s = S(&bar);
    s.0(123, 123); // <- E0057 and struct field
    F(1, 2); // <- E0057 and const item
}

fn bar(x: i32) {
    println!("{}", x);
}
```

Just notice, don't insist on supporting these cases in this PR

**Review by Undin — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -1257,6 +1263,15 @@ private fun RsCallExpr.expectedParamsCount(): Pair<Int, Boolean>? {
...
+            val type = el.type
+            // If the call is to a closure
+            if (type is TyFunction) {
```
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:1264]: @zeroeightysix let's use some heuristic here to detect only `E0057`.
For example, let's support only case `let el = closure`, i.e. when parent of `el` is `RsLetDecl` and the corresponding expression is `RsLambdaExpr`.
And after #6391, replace the heuristic with a more generic implementation

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 3]
```diff
@@ -1244,27 +1246,36 @@ private fun AnnotationSession.fileDuplicatesMap(): MutableMap<PsiElement, Map<Na
...
     }
 }
 
-private fun RsMethodCall.expectedParamsCount(): Pair<Int, Boolean>? {
+private fun RsMethodCall.expectedParamsCount(): Pair<Int, FunctionType>? {
     val fn = reference.resolve() as? RsFunction ?: return null
-    return fn.valueParameterList?.valueParameterList?.size?.let { Pair(it, fn.isVariadic) }
-        .takeIf { fn.owner.isInherentImpl }
+    return fn.valueParameterList?.valueParameterList?.size?.let {
+        Pair(it, if (fn.isVariadic) FunctionType.VARIADIC_FUNCTION else FunctionType.FUNCTION)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:1277]: nit: let's extract `if (fn.isVariadic) FunctionType.VARIADIC_FUNCTION else FunctionType.FUNCTION)` into variable

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (zeroeightysix) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:1264]: I'm not sure I follow.

*E0057* only occurs in calls, right? How can this diagnostic detect E0057 in a `let el = closure` statement?

If you mean detecting the following:
```rust
let f = || ();

f(5); // <-- E0057 here
```
I'm not sure how I would inspect the declaration of `f`.

[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:1269]: > E0057 only occurs in calls, right?

Yep

> How can this diagnostic detect E0057 in a `let el = closure` statement?
>
> If you mean detecting the following:
> ```rust
> let f = || ();
>
> f(5); // <-- E0057 here
> ```
> I'm not sure how I would inspect the declaration of `f`.

You already have `el` variable that represents `f` in the described case and you can check all necessary nodes of AST here. For example, something like
```kotlin
// TODO: replace with more generic solution
// when https://github.com/intellij-rust/intellij-rust/issues/6391 will be implemented
if (type is TyFunction) {
    val letDecl = el.parent?.parent as? RsLetDecl
    if (letDecl?.expr is RsLambdaExpr) type.paramTypes.size to FunctionType.CLOSURE else null
} else {
    null
}
```
Of course, it won't detect all possible cases, for example, it won't catch error if you write
```rust
fn main() {
    let a = |x: i32| x;
    let b = a;
    b(1, 2);
}
```

but as a temporary solution I think it's ok

**Review by Undin — APPROVED** [review]
LGTM
Thanks!
Could you squash "fixing" commits into main ones, please?

**Review by Undin — APPROVED** [review]
Thanks!
bors r+

**Pull request merged**